### ISSUE TITLE
Account for metered conncetions.

### ISSUE THREAD

**Issue by jtagcat** [creativeprojects/resticprofile#77]
https://github.com/restic/restic/issues/2353

tldr there are 4 states of metered based on
 - did the system detect the state
 - metered or not
resulting in metered-confirmed, metered-guess, unmetered-confirmed, unmetered-unconfirmed (default, very common)

Not sure about MacOS, but I'd guess there as well, OS integration is possible. Setting network (AP) names in resticprofile config doesn't sound sane.

***

 - don't run on metered connections
 - after a period of not being backed up, do also back up on metered
 - run with 1 thread (in case of roaming / interruptions (by user)), maybe bwlimit?

**Comment by creativeprojects** [comment]
It is a good idea indeed, but as I was looking around at how to get this information, it looks like it's almost impossible to detect. Most OS don't advertise it very well, and also ... they're kind of guessing anyway.

I wouldn't want to give a false sense of security where the user will end-up with thousands of _[insert your own currency here]_ bill because it wasn't detected properly.

**Comment by TilBlechschmidt** [comment]
Even though this is an old issue, I'd like to add an idea: **Filter by wireless network!**

At least for me, the only metered connection I use is my phone's WiFi hotspot which has a known SSID. So by listing this in the config I could pretty reliably prevent backups on metered connections.

This has the added benefit of (likely) not giving a false sense of security since a config option named `schedule-ignore-on-wireless-network` does not imply any guarantees beyond detecting certain wireless networks (which should ideally be testable and/or throw an error if it is not possible to detect when setting the schedule – not sure if there is a daemon independent way of detecting it on linux but on macOS/Windows it should be pretty reliable I'd imagine).

**Comment by creativeprojects** [comment]
You could still start a backup on an unmetered connection, and continue on a metered connection.

The backup will keep going if the interruption in the middle is short enough to be covered the retry mechanism of restic 🤔

**Comment by TilBlechschmidt** [comment]
Good point, would it be possible to abort a running backup if that happens (i.e. by ~nicely~ killing the process)? If I'm not mistaken restic itself is capable of dealing with that.

**Comment by creativeprojects** [comment]
yes it would be possible to keep checking in the background and send a termination signal to restic

**Comment by RazielleS** [comment]
I have attempted to do something like this for my laptop, where I want daily or possibly hourly backups for some things.
It should just not start while I'm traveling. Often just opening my laptop on the train, which would start a delayed backup almost immediately on a flaky connection.
So I set it up like this:

`profiles.d/default.yaml`
```yaml
default:
  systemd-drop-in-files:
    - "systemd/99-check-metered-connection.conf"
```

`systemd/99-check-metered-connection.conf`
```ini
[Service]
ExecCondition=/etc/resticprofile/scripts/check-metered-connection.sh
```

`scripts/check-metered-connection.sh`
```bash
#! /usr/bin/bash

metered_status=$(dbus-send --system --print-reply=literal \
        --system --dest=org.freedesktop.NetworkManager \
          /org/freedesktop/NetworkManager \
          org.freedesktop.DBus.Properties.Get \
          string:org.freedesktop.NetworkManager string:Metered \
          | grep -o ".$")

if [[ $metered_status =~ (1|3) ]]; then
  echo Current connection is metered
  exit 1
else 
  exit 0
fi
```

Now at least it won't start a backup when connected to my mobile hotspot.
I'm still looking for a way to actually delay the backup to a later time or watching for a connection change, instead of just preventing it.

EDIT:
I just stumbled over something interesting in the arch wiki:
```
systemd-run --on-active="12h 30m" --unit someunit.service
```

So it might be possible to reschedule the unit in the `ExecCondition` with this.

**Comment by creativeprojects** [comment]
I didn't realise systemd had this kind of information available now. I need to look it up in more details.

The way I do it with my laptop is very simple: just don't start any backup on battery. When I'm plugged in, I'll be either at home or in a shared workspace. I don't usually plug it in on the train. That works very well for any recent laptop with a battery in good condition 😉 

Anyway, I'm gradually changing my mind and I'm now considering adding a local scheduler to resticprofile, to make things easier with all the missing features of the default OS schedulers 🤔

**Comment by RazielleS** [comment]
Well, it's not really systemd which has this information. It would only work on machines which use NetworkManager. Of course the script could probably be adapted to be more versatile if needed.

**Comment by Mathis-Z** [comment]
Maybe the user could specify a "condition command" that is evaluated before a backup (or also during it to abort if necessary). If the command returns a non-empty result, the backup is started, otherwise it is skipped (the exit code could also be used). On Manjaro, a condition command that allowlists a single SSID could look like this: `nmcli -t -f ACTIVE,SSID dev wifi | grep "yes:TP-Link_3CC2"`

I think this would be simple to implement but also useful for stuff other than filtering for SSIDs. Additionally, it offloads the risk of any misconfiguration to the user ;)
The resticprofile documentation could also list some useful example commands for a selection of major Linux distros and Windows/macOS.

**Comment by creativeprojects** [comment]
This is a very good idea; also it shouldn't be too difficult to add (to the already rather complicated rules) 👍🏻

**Comment by creativeprojects** [comment]
Thinking of which, you can already do something similar with the `run-before` commands; if it returns a non zero exit status, it cancels the run 🤔
